### PR TITLE
Standardizing all actions

### PR DESCRIPTION
**Pull request by MatthewZMSU**
# Description

We had several actions whose responses were quite different. So we had to process their responses differently. Now All actions are the same.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

# How Has This Been Tested?
Tested it locally with several Scrapy spiders on my PC.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: helpers/utils.js [hunk 1]
```diff
@@ -177,12 +189,7 @@ exports.performAction = async function performAction(request, action) {
         const pageId = page.target()._targetId;
 
         try {
-            const response = await action(page, request);
-            return {
-                ...response,
-                contextId,
-                pageId,
-            };
+            return await action(page, request);
```
Comment (mxsnq) [inline helpers/utils.js:192]: Maybe we can perform ids assignment and closePage handling for all actions here?
`formResponse` can just collect html and cookies after wait (it should be renamed then)

File: routes/mhtml.js [hunk 2]
```diff
@@ -3,11 +3,14 @@ const utils = require('../helpers/utils');
 const router = express.Router();
 
 
-async function captureSnapshot(page) {
+async function captureSnapshot(page) {  // TODO: actually, we have ...(page, request) interface
```
Comment (mxsnq) [inline routes/mhtml.js:6]: it should work as is, but we can add it anyway

File: helpers/utils.js (hunk 1 above)
Comment (MatthewZMSU) [reply helpers/utils.js:192]: Yeah, that's what I'm thinikng about. `formResponse` does too many actions at once. Probably, we should divide it into several parts.
After this, `performAction` would close the page and get Ids.
I will deal with that.

File: routes/recaptcha_solver.js [hunk 3]
```diff
@@ -23,21 +23,19 @@ const DEFAULT_TIMEOUT = 1000;  // 1 second
...
         recaptcha_data = await page.solveRecaptchas();
-    }
-    else {
+    } else {
         recaptcha_data = await page.findRecaptchas();
     }
 
     const waitOptions = request.body.waitOptions || { timeout: DEFAULT_TIMEOUT };
-    const close_page = request.query.closePage ||
+    request.query.closePage = request.query.closePage ||
```
Comment (mxsnq) [inline routes/recaptcha_solver.js:35]: I'd like to consider request query immutable (even though it is not).
What do you think of closing the page here and adding page.isClosed() check in utils.performAction:
```javascript
        if (request.query.closePage && !page.isClosed()) {
            await page.close();
        }
        if (page.isClosed()) {
            delete response.pageId;
        }
```
That also affects closing pages in js actions.

**Review by mxsnq — APPROVED** [review]
Looks great

File: routes/recaptcha_solver.js (hunk 3 above)
Comment (MatthewZMSU) [reply routes/recaptcha_solver.js]: I'm OK with this.
This still has flexibility I wanted in the scrapy-puppeteer-service.
Tested changes with local spiders.

File: routes/recaptcha_solver.js [hunk 4]
```diff
@@ -23,21 +23,22 @@ const DEFAULT_TIMEOUT = 1000;  // 1 second
...
+    } else {
         recaptcha_data = await page.findRecaptchas();
     }
 
     const waitOptions = request.body.waitOptions || { timeout: DEFAULT_TIMEOUT };
-    const close_page = request.query.closePage ||
-        (request.body.close_on_empty && recaptcha_data['captchas'].length === 0)
+    if (request.query.closePage ||
+        (request.body.close_on_empty && recaptcha_data['captchas'].length === 0)) {
+        await page.close();
```
Comment (mxsnq) [inline routes/recaptcha_solver.js:37]: should be after `getContents`

Comment (MatthewZMSU) [reply routes/recaptcha_solver.js:37]: You are right!
It's quite strange that I did not get any bad response during testing.
I will look through my recaptcha spiders.

**Pull request merged**